client/components/title-format-editor/parser.js

```diff
@@ -0,0 +1,246 @@
+import {
+	compact,
+	flowRight as compose,
+	get,
+	has,
+	head,
+	invert,
+	mapValues,
+	matchesProperty,
+	reduce,
+	some,
+} from 'lodash';
+import {
+	convertFromRaw,
+	convertToRaw,
+} from 'draft-js';
+
+/*
+ * The functions in this file convert between the
+ * Calypso-native representation of title formats
+ * and its translation into the structure for the
+ * draft-js editor.
+ *
+ * Custom block titles consist of a list of zero
+ * or more "pieces," each of which is either a
+ * predefined "chip" or a sequence of arbitrary
+ * text.
+ *
+ * "Chips" are placeholders for dynamic content
+ * replaced at runtime, such as a site's name or
+ * a post's title.
+ *
+ * <Calypso-native title format>
+ * [
+ *   { type: 'postTitle' },
+ *   { type: 'string', value: ' | ' },
+ *   { type: 'siteName' }
+ * ]
+ *
+ * The draft-js editor however operates on a list
+ * of blocks containing plain-text and associated
+ * meta information about the characters within
+ * each block. It is domain-specific to rich text
+ * and not to our problems at hand. We must map
+ * the title formats into something the editor can
+ * use to render and edit them. See the draft-js
+ * documentation for understanding its data model.
+ *
+ * draft-js translation of above example
+ * (some data omitted to clarify the example)
+ *
+ * <RawDraftContentState>
+ * {
+ *   blocks: [
+ *     0: {
+ *       text: 'Post Title | Site Name',
+ *       type: 'unstyled',
+ *       entityRanges: [
+ *         { key: 0, offset: 0, length: 10 },
+ *         { key: 1, offset: 13, length: 9 }
+ *       ]
+ *     }
+ *   ],
+ *   entityMap: {
+ *     0: { type: 'TOKEN', mutability: 'IMMUTABLE', data: { name: 'postTitle' } },
+ *     1: { type: 'TOKEN', mutability: 'IMMUTABLE', data: { name: 'siteName' } }
+ *   }
+ * }
+ *
+ * The challenge then is to provide an effective
+ * mapping between these formats. Notice that in
+ * the editor we are showing the translations of
+ * the chip names since that is what it will
+ * render to the browser. However, the native name
+ * is stored as metadata for the range of characters
+ * that represent the chip.
+ *
+ */
+
+/**
+ * Returns whether or not at least one arg is truthy
+ *
+ * @param {bool} args pieces to 'or' together
+ * @returns {bool} result of 'or'ing all the args
+ */
+const or = ( ...args ) => some( args );
+
+/**
+ * Converts from ContentState to native Calypso format list
+ *
+ * Since the editor constrains us to a single block, we can
+ * safely assume that we only need to convert the first one
+ *
+ * @param {ContentState} content Content of editor
+ * @returns {Array} title format
+ */
+export const fromEditor = content => {
```

Comment:
How about some unit tests for the parser functions? We can do this as a follow up.